packages/components-kit/src/components/ComponentForm/ComponentForm.component.js

```diff
@@ -0,0 +1,191 @@
+		this.getUISpec = this.getUISpec.bind(this);
+		this.setupTrigger = this.setupTrigger.bind(this);
+		this.getTriggers = this.getTriggers.bind(this);
+		this.setupTrigger(props);
+	}
+
+	componentDidUpdate(props) {
+		if (
+			props.triggerURL !== this.props.triggerURL ||
+			props.customTriggers !== this.props.customTriggers
+		) {
+			this.trigger = this.setupTrigger(this.props);
+		}
+		if (this.props.definitionURL !== props.definitionURL) {
+			this.props.dispatch({
+				type: TCompForm.ON_DEFINITION_URL_CHANGED,
+				...this.props,
+			});
+		}
+	}
+
+	onChange(event, data) {
+		event.persist();
+		if (!this.props.state.get('dirty')) {
+			this.props.setState({ dirty: true });
+		}
+		const properties = data.properties;
+		if (data.schema.titleMap && data.value) {
+			// Here we add a field side by side with the value
+			// to keep the title associated to the value
+			const info = data.schema.titleMap.find(titleMap => titleMap.value === data.value);
+			let currentProp = properties;
+			let currentKey;
+			data.schema.key.forEach((key, index) => {
+				if (index !== data.schema.key.length - 1) {
+					currentProp = currentProp[key];
+				} else {
+					currentKey = key;
+				}
+			});
+			currentProp[`$${currentKey}_name`] = info.name;
+		}
+		this.setState({ properties });
+		if (this.props.dispatchOnChange) {
+			this.props.dispatch({
+				type: TCompForm.ON_CHANGE,
+				event: {
+					type: 'onChange',
+					component: 'TCompForm',
+					props: this.props,
+					state: this.state,
+					source: event,
+				},
+				data,
+				properties,
+				uiSpec: this.getUISpec(),
+			});
+		}
+	}
+
+	onTrigger(event, payload) {
+		return this.trigger(event, payload).then(data => {
+			if (data.properties) {
+				this.setState({ properties: data.properties });
+			}
+			if (data.errors || data.jsonSchema || data.uiSchema) {
+				this.props.setState(data);
+			}
+			this.props.dispatch({
+				type: TCompForm.ON_TRIGGER,
+				event: {
+					type: 'onTrigger',
+					component: ComponentForm,
+					componentId: this.props.componentId,
+					props: this.props,
+					state: this.state,
+				},
+				data,
+				uiSpec: this.getUISpec(),
+			});
+		});
+	}
+
+	onSubmit(event, data) {
+		event.persist();
+		if (this.props.onSubmit) {
+			this.props.onSubmit(event, data);
+		}
+		this.props.dispatch({
+			type: TCompForm.ON_SUBMIT,
+			event,
+			...this.getUISpec(),
+			properties: data,
+		});
+	}
+
+	setupTrigger(props) {
+		this.trigger = kit.createTriggers({
+			url: props.triggerURL,
+			customRegistry: this.getTriggers(),
+		});
+	}
+
+	getTriggers() {
+		return {
+			reloadForm: ({ body }) => ({
+				...body,
+				properties: { _datasetMetadata: this.state.properties._datasetMetadata },
+			}),
+		};
+	}
+
+	getUISpec() {
+		const spec = { properties: this.state.properties };
+		const jsonSchema = this.props.state.get('jsonSchema');
+		const uiSchema = this.props.state.get('uiSchema');
+		if (jsonSchema) {
+			spec.jsonSchema = jsonSchema.toJS();
```

Comment:
why to JS ? can't we take this opportunity to avoid those ?

Reply:
UIForm use json-schema-form-core (jsfc) that process the json and ui schemas, as js objects.
We need to pass js objects, not immutablejs objects.
I just changed that to memoize the last result for jsonSchema, uiSchema and errors